docs/developer-docs/smart-contracts/best-practices/idempotency.mdx

```diff
@@ -0,0 +1,90 @@
+
+## Other approaches to safe retries
+
+In absence of idempotent endpoints, or even as an addition to them, clients may be able to use other endpoints to make their retries safe.
+
+### Queryable call results
+
+If the canister, in addition to the update endpoint, also exposes a query that can inform the user of the result of the update, the client can also use this for safe retries as follows:
+
+1. Attempt to perform the update
+1. If the result of the update is unknown (e.g., not present in the ingress history any more), query the call result endpoint to determine whether the call was applied or not. Moreover, one needs to ensure that the previously sent call call cannot be applied in the future. If both of these are true, the call might be retried, or safely reported as failed.
+
+In practice, this pattern may be more complicated. For example, the ICP ledger exposes a `query_blocks` method that can be used to implement the above pattern for transfers initiated as ingress messages:
+
+1. Call the `query_blocks` method on the ledger to determine what the last block (as specified in the `chain_length` field of the response) currently is. Let's call this `last_block`.
+1. Attempt to perform a transfer. This ingress message includes an `ingress_expiry` field.
+1. If the result of the transfer is unknown, call the `read_state` endpoint on the ledger canister, to obtain the `/time` branch of the system state tree. Repeat this until the reported time exceeds the `ingress_expiry` time. This ensures that the transfer will not be applied at a later point.
+1. Call the `query_blocks` method on the ledger again to retrieve all ledger blocks since `last_block`, and check that the `timestamp` also exceeds the `ingress_expiry` time. Then, scan through the returned blocks to determine whether the transaction has been included or not.
+
+### 2-step transfers
+
+Another approach applicable to ledgers (such as ICRC-1 or ICP) is to perform transfers in two steps:
+
+1. First, transfer the money to an intermediate subaccount of the sender that's specific to this transaction. For example, if the transaction has a unique ID, the client can hash the ID to obtain a subaccount. The transferred amount should be the desired amount increased by the ledger transaction fee.
+1. If the result of the above transfer is unknown, query the balance of the transaction-specific subaccount. Like in the ["queryable call result"](#queryable-call-results) approach, this should be repeated until the `timestamp` accompanying the response exceeds the `ingress_expiry`. If the balance is 0, the transaction can safely be reported as failed, or retried (starting from step 1). If the balance is at least the expected balance, one can proceed.
```

Comment:
```suggestion
1. If the result of the above transfer is unknown, query the balance of the transaction-specific subaccount. Like in the ["queryable call result"](#queryable-call-results) approach, this should be repeated until the `timestamp` accompanying the response exceeds the `ingress_expiry`. If the balance is 0, the transaction can safely be reported as failed, or it can be retried (starting from step 1). If the balance is at least the expected balance, one can proceed.
```